File: src/test/java/im/tox/tox4j/callbacks/FriendTypingCallbackTest.java
package im.tox.tox4j.callbacks;

import im.tox.tox4j.AliceBobTestBase;
import im.tox.tox4j.ToxCore;
import im.tox.tox4j.ToxCoreImpl;
import im.tox.tox4j.ToxOptions;
import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.enums.ToxConnection;
import im.tox.tox4j.exceptions.ToxException;
import im.tox.tox4j.exceptions.ToxNewException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;

public class FriendTypingCallbackTest extends AliceBobTestBase {

    @NotNull
    @Override
    protected ChatClient newAlice() {
        return new Client();
    }

    private static class Client extends ChatClient {

        private boolean initial = true;

        private void setTyping(final int friendNumber, final boolean isTyping) {
            addTask(new Task() {
                @Override
                public void perform(@NotNull ToxCore tox) throws ToxException {
                    tox.setTyping(friendNumber, isTyping);
                }
            });
        }

        public void friendConnectionStatus(final int friendNumber, @NotNull ToxConnection connection) {
            if (connection != ToxConnection.NONE) {
                debug("is now connected to friend " + friendNumber);
                // Alice starts typing.
                if (isAlice()) {
                    setTyping(friendNumber, true);
                }
            }
        }

        @Override
        public void friendTyping(int friendNumber, boolean isTyping) {
            if (initial) {
                // We get the initial message that both are not typing.
                assertFalse(isTyping);
                initial = false;
                return;
            }
            if (isTyping) {
                debug("friend is now typing");
            } else {
                debug("friend stopped typing");
            }
            assertEquals(FRIEND_NUMBER, friendNumber);
            if (isBob() && isTyping) {
                // Alice started typing to Bob. Now Bob starts typing.
                setTyping(friendNumber, true);
            }
            if (isAlice() && isTyping) {
                // Bob started typing. This makes Alice stop typing.
                setTyping(friendNumber, false);
            }
            if (isBob() && !isTyping) {
                // Alice stopped typing. Now Bob also stops.
                setTyping(friendNumber, false);
                // Bob is done.
                finish();
            }
            if (isAlice() && isTyping) {
                // Bob also stopped typing. Alice is done.
                finish();
            }
        }

    }

}

File: src/test/java/im/tox/tox4j/core/callbacks/FriendTypingCallbackTest.scala
package im.tox.tox4j.core.callbacks

import im.tox.tox4j.core.enums.ToxConnection
import im.tox.tox4j.testing.autotest.{ AliceBobTest, AliceBobTestBase, ChatClient }
import org.junit.Assert.{ assertEquals, assertFalse }

class FriendTypingCallbackTest extends AliceBobTest {

  protected override def newAlice(name: String, expectedFriendName: String) = new ChatClient(name, expectedFriendName) {

    private var initial = true

    private def setTyping(friendNumber: Int, isTyping: Boolean): Unit = {
      addTask { tox =>
        tox.setTyping(friendNumber, isTyping)
      }
    }

    override def friendConnectionStatus(friendNumber: Int, connection: ToxConnection): Unit = {
      if (connection != ToxConnection.NONE) {
        debug("is now connected to friend " + friendNumber)
        if (isAlice) {
          setTyping(friendNumber, isTyping = true)
        }
      }
    }

    override def friendTyping(friendNumber: Int, isTyping: Boolean): Unit = {
      if (initial) {
        assertFalse(isTyping)
        initial = false
      } else {
        if (isTyping) {
          debug("friend is now typing")
        } else {
          debug("friend stopped typing")
        }
        assertEquals(AliceBobTestBase.FRIEND_NUMBER, friendNumber)
        if (isBob) {
          if (isTyping) {
            setTyping(friendNumber, isTyping = true)
          } else {
            setTyping(friendNumber, isTyping = false)
            finish()
          }
        } else {
          if (isTyping) {
            setTyping(friendNumber, isTyping = false)
            finish()
          }
        }
      }
    }
  }

}
